### PR TITLE
Add comprehensive test coverage for proxy and MITM components

### PR DESCRIPTION
**Pull request by laplaque**
## Summary
This PR significantly expands test coverage for the proxy and MITM packages by adding 100+ new test cases covering request/response handling, compression, anonymization, SSRF protection, and edge cases. It also introduces a delta coverage check script to enforce coverage thresholds on changed code.

### Test Coverage Additions

**Proxy Package (`internal/proxy/proxy_test.go`)**
- Added helper functions for test setup (`newTestProxyServer`, `newTestProxyServerAllowLocal`, `backendHostPort`)
- Utility function tests: `hashRemoteAddr`, `toSet`, `removeHopByHop`, `copyHeader`
- Response decompression tests: gzip, deflate, identity, empty, unsupported, and invalid data handling
- Streaming response detection tests
- Response body deanonymization tests covering non-streaming, streaming, gzip-encoded, and error cases
- MITM metrics recording tests
- HTTP request handling tests: passthrough, AI anonymization, auth passthrough, private host blocking, upstream errors
- CONNECT tunnel tests: private host blocking, missing CA handling
- Request anonymization tests: nil body, zero content length, valid body, read errors
- Forward request tests with gzip response handling
- CA file loading tests with invalid paths
- SSRF protection edge cases: no port, localhost resolution
- Integration tests for MITM request serving with both anonymization and auth passthrough

**MITM Package (`internal/mitm/mitm_test.go`)**
- HTTP/1.1 connection handling test with TLS handshake
- Bad TLS handshake error handling
- CA loading edge cases: invalid key PEM, invalid cert DER, invalid key DER
- Certificate cache eviction test
- HTTP/2 protocol negotiation test
- Expired certificate regeneration test
- CA generation error paths: bad cert and key file paths

**Management Package (`internal/management/management_test.go`)**
- Domain removal with wrong HTTP method
- Domain removal with empty body validation
- Metrics endpoint availability tests (disabled and enabled states)
- Domain registry persistence tests: no persist path, atomic writes
- Authentication tests: malformed bearer token handling

### CI/CD Improvements
- Added `delta-coverage.sh` script to enforce coverage thresholds on changed code in pull requests
- Integrated delta coverage check into GitHub Actions CI workflow
- Script validates that all functions in modified files meet the specified coverage threshold (default 95%)

## Implementation Details

- Test helpers use `httptest` servers to simulate real HTTP backends while bypassing SSRF protections for testing
- Compression tests verify both successful decompression and error handling for gzip/deflate formats
- MITM tests use `net.Pipe()` for in-process TLS connection testing without external network calls
- Delta coverage script uses `go tool cover` output parsing and `git diff` to identify changed files and validate coverage
- All new tests follow existing patterns and include proper cleanup with `t.Cleanup()`

https://claude.ai/code/session_0139or3GJtSP937fLz42zeKU

### CODE REVIEW

File: internal/proxy/proxy_test.go [hunk 1]
```diff
@@ -412,3 +416,862 @@ func TestProcessMITMRequestBody_SuccessfulAnonymization(t *testing.T) {
...
+	// CA is nil, so even AI domain goes to opaque tunnel
+	req := httptest.NewRequestWithContext(context.Background(), http.MethodConnect, "http://api.openai.com:443", nil)
+	req.Host = "api.openai.com:443"
+
+	w := httptest.NewRecorder()
+	srv.handleTunnel(w, req)
+
+	// Should try opaque tunnel (which will fail to dial, returning 502)
+	if w.Code != http.StatusBadGateway {
+		t.Errorf("expected 502, got %d", w.Code)
```
Comment (laplaque) [inline internal/proxy/proxy_test.go]: **CRITICAL:** Test expects 502 but gets 500 — this is the CI failure.

When `ca == nil`, `handleTunnel` falls through to `handleOpaqueTunnel`. That function dials `api.openai.com:443` successfully, then calls `w.(http.Hijacker)` on `httptest.NewRecorder()`, which doesn't implement `Hijacker`. The code correctly returns 500 (`http.StatusInternalServerError`) at `proxy.go:388`.

The test expectation is wrong, not the production code.

```go
// Fix: match the actual behavior
if w.Code != http.StatusInternalServerError {
    t.Errorf("expected 500 for missing hijacker, got %d", w.Code)
}
```

Alternatively, consider that this test depends on external DNS resolution of `api.openai.com` — it will behave differently in network-isolated CI environments. A more robust approach would mock the dialer to fail immediately, which *would* produce 502.

File: internal/proxy/proxy_test.go [hunk 2]
```diff
@@ -412,3 +416,862 @@ func TestProcessMITMRequestBody_SuccessfulAnonymization(t *testing.T) {
...
+		t.Errorf("expected 2 X-Multi values, got %d", len(vals))
+	}
+}
+
+// --- decompressResponse ---
+
+func TestDecompressResponse_Gzip(t *testing.T) {
+	var buf bytes.Buffer
+	gw := gzip.NewWriter(&buf)
+	gw.Write([]byte("hello gzip")) //nolint:errcheck
```
Comment (laplaque) [inline internal/proxy/proxy_test.go]: **MEDIUM:** `//nolint:errcheck` without justification. These `Write` and `Close` calls on gzip/flate writers can fail, and in test setup code a failure means the test fixture is broken.

While it's common to suppress error checks in test helpers, each suppression should have a brief justification. Better: check the error and call `t.Fatal`.

```go
if _, err := gw.Write([]byte("hello gzip")); err != nil {
    t.Fatalf("gzip write: %v", err)
}
if err := gw.Close(); err != nil {
    t.Fatalf("gzip close: %v", err)
}
```

This pattern repeats at lines 588-589, 750-751, 1017-1018, 1025, and 1133.

File: .github/scripts/delta-coverage.sh [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+    below=$(awk "BEGIN {print ($pct < $THRESHOLD) ? 1 : 0}")
+    if [ "$below" -eq 1 ]; then
+      echo "FAIL: ${line}"
+      failed=$((failed + 1))
+    fi
+  done < <(echo "$cover_output" | grep "$file" || true)
+done <<< "$changed_files"
+
+echo ""
+echo "Checked ${checked} functions in ${changed_files_count:=$(echo "$changed_files" | wc -l)} changed files."
```
Comment (laplaque) [inline .github/scripts/delta-coverage.sh]: **HIGH:** Bug — `changed_files_count` is used in the `echo` but was never assigned before this line. The `:=` default assignment inside `$(...)` sets it inline, but this is fragile and confusing. If `changed_files` is empty (early exit already happened), this is unreachable, but the variable should still be set explicitly for clarity.

```bash
# After the early exit block, add:
changed_files_count=$(echo "$changed_files" | wc -l)

# Then simplify line 68 to:
echo "Checked ${checked} functions in ${changed_files_count} changed files."
```

File: .github/scripts/delta-coverage.sh [hunk 4]
```diff
@@ -0,0 +1,75 @@
...
+# Get function-level coverage
+cover_output=$(go tool cover -func="$COVERAGE_FILE")
+
+failed=0
+checked=0
+
+while IFS= read -r file; do
+  # go tool cover uses module-relative paths; strip the leading directory
+  # to match against the file paths from git diff.
+  # The cover output format is: <module-path>/<file>:<line>:\t<func>\t<coverage>
```
Comment (laplaque) [inline .github/scripts/delta-coverage.sh]: **MEDIUM:** The `grep "$file"` match is a substring match against `go tool cover` output. If a file path like `proxy.go` appears as a substring of another file (e.g., `reverse_proxy.go`), this will produce false matches and potentially report coverage failures for the wrong file.

Use a more precise match:
```bash
done < <(echo "$cover_output" | grep -F "/$file:" || true)
```
The trailing `:` (from the `file:line:` format) plus `-F` for literal matching prevents substring collisions.

File: internal/management/management_test.go [hunk 5]
```diff
@@ -301,3 +302,106 @@ func TestRemoveDomain_InvalidDomain(t *testing.T) {
...
+
+	if w.Code != http.StatusOK {
+		t.Errorf("expected 200 for metrics, got %d", w.Code)
+	}
+	if ct := w.Header().Get("Content-Type"); ct != "application/json" {
+		t.Errorf("expected application/json, got %q", ct)
+	}
+}
+
+func TestDomainRegistry_PersistNoPersistPath(_ *testing.T) {
```
Comment (laplaque) [inline internal/management/management_test.go]: **LOW:** Unused `*testing.T` parameter (named `_`). While this compiles fine, it means the test can't call `t.Fatal`/`t.Error` if the "should not panic" assumption is violated. Consider using `t` and adding a `recover()` assertion, or at minimum name the parameter for consistency.

```go
func TestDomainRegistry_PersistNoPersistPath(t *testing.T) {
    t.Helper()
    // ...
}
```

Same issue at `proxy_test.go:784` (`TestRecordMITMMetrics_NilMetrics`).

File: internal/proxy/proxy_test.go [hunk 6]
```diff
@@ -412,3 +416,862 @@ func TestProcessMITMRequestBody_SuccessfulAnonymization(t *testing.T) {
...
+		t.Errorf("expected 'backend response', got %q", w.Body.String())
+	}
+}
+
+func TestServeHTTP_HTTP_AIAnonymization(t *testing.T) {
+	backend := httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
+		body, _ := io.ReadAll(r.Body)
+		w.Header().Set("Content-Type", "application/json")
+		w.WriteHeader(http.StatusOK)
+		w.Write(body) //nolint:errcheck
```
Comment (laplaque) [inline internal/proxy/proxy_test.go]: **MEDIUM:** `//nolint:errcheck` on `w.Write(body)` inside an `http.HandlerFunc` test backend. If the write fails, the test will pass with incorrect data flowing through the proxy, masking bugs. In test backends, prefer checking the error:

```go
if _, err := w.Write(body); err != nil {
    http.Error(w, err.Error(), http.StatusInternalServerError)
    return
}
```

File: internal/proxy/proxy_test.go [hunk 7]
```diff
@@ -412,3 +416,862 @@ func TestProcessMITMRequestBody_SuccessfulAnonymization(t *testing.T) {
...
+
+	if w.Code != http.StatusForbidden {
+		t.Errorf("expected 403, got %d", w.Code)
+	}
+}
+
+func TestHandleTunnel_AIDomainWithoutCA(t *testing.T) {
+	srv := newTestProxyServer(t)
+	// CA is nil, so even AI domain goes to opaque tunnel
+	req := httptest.NewRequestWithContext(context.Background(), http.MethodConnect, "http://api.openai.com:443", nil)
```
Comment (laplaque) [inline internal/proxy/proxy_test.go]: **HIGH:** This test depends on successfully dialing `api.openai.com:443` from the test environment. This is an external network dependency that makes the test:
1. Flaky in network-restricted CI environments
2. Slow (real TCP handshake)
3. Non-deterministic (DNS/routing changes)

Inject a mock dialer that either fails (to test the 502 path) or succeeds (to test the hijacker path). The `newTestProxyServerAllowLocal` pattern already shows how to override `srv.dialContext`.

File: internal/mitm/mitm_test.go [hunk 8]
```diff
@@ -354,3 +357,263 @@ func TestSingleConnListener_Addr(t *testing.T) {
...
+	dir := t.TempDir()
+	keyFile := filepath.Join(dir, "key.pem")
+
+	badKeyPEM := "-----BEGIN RSA PRIVATE KEY-----\nYmFkZGF0YQ==\n-----END RSA PRIVATE KEY-----\n"
+	if err := os.WriteFile(keyFile, []byte(badKeyPEM), 0600); err != nil {
+		t.Fatal(err)
+	}
+	_, err := LoadCA(certFile, keyFile)
+	if err == nil {
+		t.Error("expected error for invalid key DER")
```
Comment (laplaque) [inline internal/mitm/mitm_test.go:491]: **LOW:** The `TestHandleConn_H2` test has a best-effort approach — it accepts the handler not being called before close as OK for coverage. The comment `// H2 handler may not have been called before close; that's OK for coverage` is honest, but it means this test doesn't actually verify HTTP/2 request handling works. Consider using `golang.org/x/net/http2` client transport over the TLS conn to make a proper H2 request and assert the response.

File: .github/workflows/ci.yml [hunk 9]
```diff
@@ -48,6 +48,12 @@ jobs:
       - name: Display coverage
         run: go tool cover -func=coverage.out
 
+      - name: Delta coverage check
+        if: github.event_name == 'pull_request'
+        run: |
```
Comment (laplaque) [inline .github/workflows/ci.yml:53]: **LOW:** `git fetch origin main --depth=1` fetches only a shallow commit. If the PR branch diverged significantly from main, the `git diff --name-only` in `delta-coverage.sh` may not have the full merge base available, causing it to report incorrect changed files (or fail). Consider using `--unshallow` or fetching enough depth for the merge base:

```yaml
run: |
  git fetch origin main --depth=50
  bash .github/scripts/delta-coverage.sh coverage.out 95.0 origin/main
```

**Review by laplaque — COMMENTED** [review]
## PR Review: Add comprehensive test coverage for proxy and MITM components

**Scope:** 1311 additions, 0 deletions, 5 files.

### Pipeline status
**BLOCKED** — `Test` job is failing. Root cause: `TestHandleTunnel_AIDomainWithoutCA` expects HTTP 502 but gets HTTP 500. When `ca == nil`, `handleTunnel` falls through to `handleOpaqueTunnel`, which dials `api.openai.com:443` (succeeds on real DNS), then attempts `w.(http.Hijacker)` on `httptest.NewRecorder` — which doesn't implement `Hijacker`, yielding 500 (Internal Server Error). The test expectation of 502 is incorrect; the actual behavior (500) is the correct response for "hijacking not supported". Fix: change the expected status to `http.StatusInternalServerError` (500), or restructure the test to use a hijack-capable recorder.

Additionally, `Build` job is skipped because `Test` failed.

### Test coverage verdict
**UNVERIFIABLE (for delta) / FAIL (for package-wide proxy)**

Package-wide numbers (not delta):
- `internal/proxy`: **83.7%** — below 85% threshold
- `internal/mitm`: **85.7%**
- `internal/management`: **86.3%**

Key uncovered functions in changed files:
- `handleMITMTunnel`: **0.0%** — critical MITM interception path with no test coverage
- `handleOpaqueTunnel`: **45.2%** — tunnel establishment path barely covered
- `handleTunnel`: **75.0%**

No delta coverage tooling runs successfully yet (the new `delta-coverage.sh` never executes because `Test` fails first). Recommend fixing the test failure so the delta coverage gate can be validated end-to-end.

### Strengths
- Excellent breadth: 100+ new test cases covering decompression, SSRF, streaming, metrics, auth, and edge cases
- Good use of `httptest` servers and `net.Pipe()` for realistic in-process testing
- The delta-coverage.sh script is a solid addition for CI enforcement
- Tests follow existing patterns with proper `t.Cleanup()`

### Verdict
**Not ready to merge.** Blockers:
1. CI `Test` job is failing (`TestHandleTunnel_AIDomainWithoutCA` wrong expected status)
2. `handleMITMTunnel` at 0% coverage — this is the core MITM interception path
3. Multiple `//nolint:errcheck` suppressions without justification in test code
4. `delta-coverage.sh` has a bug using an uninitialized variable

---
*Review by Claude*

File: internal/proxy/proxy_test.go (hunk 1 above)
Comment (laplaque) [reply internal/proxy/proxy_test.go]: Fixed in `9d2fe96`. Rewrote the test to use a private IP literal (`10.0.0.52`) registered as an AI domain, so the opaque tunnel SSRF check blocks it immediately with 403. This eliminates the external DNS dependency entirely and makes the test instant instead of timing out.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/proxy/proxy_test.go (hunk 2 above)
Comment (laplaque) [reply internal/proxy/proxy_test.go]: Fixed in `fe3b108`. Replaced all `//nolint:errcheck` suppressions in test fixtures with proper error checks calling `t.Fatalf`. Covers gzip Write/Close, flate NewWriter/Write/Close, and all `w.Write(body)` calls in test backends.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/scripts/delta-coverage.sh (hunk 3 above)
Comment (laplaque) [reply .github/scripts/delta-coverage.sh]: Fixed in `fe3b108`. Moved `changed_files_count=$(echo "$changed_files" | wc -l)` to right after the early-exit block, before usage in the echo statement.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/scripts/delta-coverage.sh (hunk 4 above)
Comment (laplaque) [reply .github/scripts/delta-coverage.sh]: Fixed in `fe3b108`. Changed to `grep -F "/$file:"` — the leading `/` plus trailing `:` (from the `file:line:` format) with `-F` literal matching prevents substring collisions.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/management/management_test.go (hunk 5 above)
Comment (laplaque) [reply internal/management/management_test.go]: Fixed in `fe3b108`. Both `TestDomainRegistry_PersistNoPersistPath` and `TestRecordMITMMetrics_NilMetrics` now accept `t *testing.T` and wrap the body in a `defer recover()` that calls `t.Errorf` on panic.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/proxy/proxy_test.go (hunk 6 above)
Comment (laplaque) [reply internal/proxy/proxy_test.go]: Fixed in `fe3b108`. All `w.Write(body)` calls in test backend handlers now check the error and return 500 on failure. Applied to the AI anonymization, MITM integration, and gzip response test backends.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/proxy/proxy_test.go (hunk 7 above)
Comment (laplaque) [reply internal/proxy/proxy_test.go]: Fixed in `9d2fe96`. Eliminated the external DNS dependency entirely by rewriting the test to use a private IP literal (`10.0.0.52`) added to `aiDomains`. The SSRF check in `handleOpaqueTunnel` blocks it immediately with 403, making the test deterministic and instant.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: internal/mitm/mitm_test.go (hunk 8 above)
Comment (laplaque) [reply internal/mitm/mitm_test.go:491]: Fixed in `fe3b108`. The test now uses `http2.Transport.NewClientConn()` to create a proper H2 client, sends a real `GET` request via `h2Conn.RoundTrip()`, and asserts the 200 response status. No more best-effort — it fully verifies HTTP/2 request handling end-to-end.

---
_Generated by [Claude Code](https://claude.ai/code)_

File: .github/workflows/ci.yml (hunk 9 above)
Comment (laplaque) [reply .github/workflows/ci.yml:53]: Fixed in `fe3b108`. Increased to `--depth=50` to ensure the merge base is available for branches that diverge significantly from main.

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by laplaque — COMMENTED** [review]
## Re-review: Add comprehensive test coverage for proxy and MITM components

**Commits reviewed:** 2 new commits since initial review (`9d2fe96`, `fe3b108`).

### Pipeline status
**PASS** — All 8 checks green: Test, Build, Lint, Security Scan, Benchmark, CodeQL, Analyze (go), Analyze (actions).

### Original findings — resolution status

| # | Severity | Finding | Status |
|---|----------|---------|--------|
| 1 | CRITICAL | `TestHandleTunnel_AIDomainWithoutCA` expects 502, gets 500 | **FIXED** — Test now uses private IP so SSRF blocks immediately (no external DNS) |
| 2 | HIGH | Test depends on external DNS (`api.openai.com`) | **FIXED** — Replaced with `10.0.0.52` private IP literal |
| 3 | HIGH | `delta-coverage.sh` uninitialized `changed_files_count` | **FIXED** — Variable assigned at line 32 before use at line 69 |
| 4 | MEDIUM | `grep "$file"` substring matching in delta-coverage.sh | **FIXED** — Now uses `grep -F "/$file:"` for precise matching |
| 5 | MEDIUM | 11x `//nolint:errcheck` without justification | **FIXED** — All removed from test code; errors now checked with `t.Fatal`. One justified suppression remains on a type assertion (`proxy_test.go:1379`) with comment |
| 6 | MEDIUM | `w.Write(body)` errcheck in test backend | **FIXED** — Errors now handled |
| 7 | LOW | Unused `*testing.T` param in two tests | **FIXED** — Both `TestDomainRegistry_PersistNoPersistPath` and `TestRecordMITMMetrics_NilMetrics` now accept `t` |
| 8 | LOW | H2 test doesn't verify actual HTTP/2 request | **FIXED** — Now uses `http2.ClientConn.RoundTrip()` for a proper H2 request and asserts the response |
| 9 | LOW | `git fetch --depth=1` in CI | **FIXED** — Now uses `--depth=50` |

### New coverage numbers

| Package | Before PR | After (commit 1) | After (commit 3) |
|---------|-----------|------------------|------------------|
| `internal/proxy` | ~31% | 83.7% | **85.7%** |
| `internal/mitm` | ~65.5% | 85.7% | **85.7%** |
| `internal/management` | ~80.6% | 86.3% | **86.3%** |

Key function coverage improvements:
- `handleMITMTunnel`: 0% → **88.2%** (was the biggest gap, now covered with hijack-capable writer test)
- `handleOpaqueTunnel`: 45.2% → 16.1% (decreased — see note below)

**Note on `handleOpaqueTunnel` regression:** Coverage dropped from 45.2% to 16.1% because the `TestHandleTunnel_AIDomainWithoutCA` fix changed the test to use a private IP (which is blocked by SSRF before reaching the opaque tunnel dial/hijack paths). This is the correct trade-off — a deterministic test that covers SSRF blocking is better than one that accidentally covered opaque tunnel internals via a flaky DNS-dependent path. Dedicated `handleOpaqueTunnel` coverage should be added in a follow-up.

### Test suite results (local)
- `go test ./...` — **all PASS** (0 failures)
- All anonymizer pack unit tests — **all PASS** (GLOBAL, US, DE, FR, NL, SECRETS, FINANCE_EU, HEALTHCARE)
- All report integration tests — **all PASS** (finance_eu, healthcare, nl, secrets, secrets_priority, cross-pattern)

### Remaining items (non-blocking)
1. `handleOpaqueTunnel` at 16.1% — recommend a follow-up PR with a test using a hijack-capable writer and local TCP backend
2. One `//nolint:errcheck` remains at `proxy_test.go:1379` on a type assertion — has justification comment, acceptable
3. `handleTunnel` at 75.0% — the MITM+auth dispatch path needs an additional test where `ca != nil` and domain is auth-only

### Verdict
**Ready to merge.** All 9 original findings are resolved. CI is fully green. All pack tests pass. The remaining coverage gaps are pre-existing or architectural and appropriate for a follow-up PR.

---
*Re-review by Claude*

**Comment by laplaque** [comment]
## Follow-up: `handleOpaqueTunnel` coverage

The fix for `TestHandleTunnel_AIDomainWithoutCA` was correct (deterministic SSRF block vs flaky external DNS), but it dropped `handleOpaqueTunnel` from 45.2% to 16.1%. Here's a concrete approach for a follow-up PR to cover the happy path:

**Pattern:** Use `net.Listen("tcp", "127.0.0.1:0")` as a local backend + a hijack-capable `ResponseWriter` (same pattern as `TestHandleMITMTunnel` from `fe3b108`). Override `srv.dialContext` via `newTestProxyServerAllowLocal` so the SSRF check doesn't block `127.0.0.1`.

```go
func TestHandleOpaqueTunnel_HappyPath(t *testing.T) {
    // 1. Start a local TCP echo server
    ln, err := net.Listen("tcp", "127.0.0.1:0")
    if err != nil {
        t.Fatal(err)
    }
    defer ln.Close()
    go func() {
        conn, err := ln.Accept()
        if err != nil { return }
        defer conn.Close()
        io.Copy(conn, conn) // echo
    }()

    // 2. Create proxy with local dial allowed
    srv := newTestProxyServerAllowLocal(t, nil, nil)

    // 3. Build a hijack-capable ResponseWriter (reuse the hijackWriter from TestHandleMITMTunnel)
    clientConn, serverConn := net.Pipe()
    hw := &hijackWriter{conn: serverConn, ResponseWriter: httptest.NewRecorder()}

    // 4. Call handleOpaqueTunnel with the local listener address
    host := ln.Addr().String()
    req := httptest.NewRequestWithContext(t.Context(), http.MethodConnect, "http://"+host, nil)
    req.Host = host

    go srv.handleOpaqueTunnel(hw, req, host)

    // 5. Send data through the tunnel and verify echo
    _, _ = clientConn.Write([]byte("ping"))
    buf := make([]byte, 4)
    clientConn.SetReadDeadline(time.Now().Add(2 * time.Second))
    n, err := clientConn.Read(buf)
    if err != nil || string(buf[:n]) != "ping" {
        t.Errorf("expected echo 'ping', got %q (err: %v)", buf[:n], err)
    }
    clientConn.Close()
}
```

This covers: dial success, hijack, `flushingCopy` in both directions, and connection cleanup. No production code changes needed.

**Constraints for Claude Code:**
- Don't touch production code — test should work with existing `handleOpaqueTunnel` signature
- Don't introduce external network dependencies (no real DNS, no public IPs)
- Keep `TestHandleTunnel_AIDomainWithoutCA` as-is — it correctly covers the SSRF-block path
- Target: `handleOpaqueTunnel` above 80%

**Comment by laplaque** [comment]
Implemented in `d6805d3`. Added both tests following the suggested pattern:

- **`TestHandleOpaqueTunnel_HappyPath`** — local TCP echo server + hijack-capable `ResponseWriter`, verifies dial → hijack → bidirectional copy → echo round-trip
- **`TestHandleOpaqueTunnel_DialFailure`** — verifies 502 when backend is unreachable

Used `localhost:<port>` instead of the raw `127.0.0.1` IP to pass `isPrivateHost` (which only checks literal IPs), with `newTestProxyServerAllowLocal` bypassing the SSRF dial-time check.

Results:
- `handleOpaqueTunnel`: 16.1% → **87.1%**
- `internal/proxy` package: 85.7% → **93.0%**
- Overall: 90.4% → **92.0%**

---
_Generated by [Claude Code](https://claude.ai/code)_

**Review by laplaque — COMMENTED** [review]
## Final review — all clear on `d6805d3`

CI: **8/8 green** (Test, Build, Lint, Security Scan, Benchmark, CodeQL, Analyze go, Analyze actions).

Local test run: **all pass**, 0 failures. Overall statement coverage: **92.0%**.

### Package coverage (final)

| Package | Coverage |
|---------|----------|
| `internal/proxy` | **93.0%** |
| `internal/anonymizer` | 96.0% |
| `internal/anonymizer/packs` | 97.6% |
| `internal/config` | 100.0% |
| `internal/metrics` | 100.0% |
| `internal/management` | 86.3% |
| `internal/logger` | 86.4% |
| `internal/mitm` | 85.7% |

### Key function coverage (proxy)

| Function | Start of PR | Final |
|----------|------------|-------|
| `handleOpaqueTunnel` | ~45% | **87.1%** |
| `handleMITMTunnel` | 0% | **88.2%** |
| `serveMITMRequest` | — | **90.9%** |
| `forward` | — | **96.0%** |
| `deanonymizeResponseBody` | — | **95.2%** |
| `decompressResponse` | — | **100.0%** |

### Pack test results (all PASS on `d6805d3`)

**Unit tests** (`internal/anonymizer/packs/`):
- DE: `TestDESteuerIDPattern`, `TestDESVNRPattern`, `TestDEKfzPattern`, `TestValidateSteuerID` (12 subcases)
- FR: `TestFRNIRPattern`, `TestFRSIRETPattern`, `TestFRSIRENPattern`, `TestValidateFRNIR` (11 subcases), `TestValidateLuhn` (9 subcases), `TestValidateSIRET` (6), `TestValidateSIREN` (6)
- NL: `TestNLBSNPattern`, `TestNLKvKPattern`, `TestValidateBSN` (8 subcases)
- US: `TestUSSSNPattern`, `TestUSPhonePattern`, `TestUSAddressPattern` (9 subcases), `TestUSIPv4Pattern`, `TestUSIPv6Pattern`, `TestValidateSSN` (10), `TestValidateUSPhone` (6)
- GLOBAL: `TestGlobalEmailPattern`, `TestGlobalAPIKeyPattern`, `TestGlobalCreditCardWithLuhn`, `TestValidateEmailLocalPart` (6)
- SECRETS: 24 pattern tests + `TestSecretsOpenAIvsStripeNoCollision`, `TestSecretsGroqvsGitHubNoCollision`, `TestSecretsTwilioTightHexRequirement`
- FINANCE_EU: `TestFINANCEEUIBANPattern`, `TestFINANCEEUSWIFTBICPattern`, `TestFINANCEEUVATPattern`, `TestValidateIBAN` (10)
- HEALTHCARE: `TestHEALTHCAREMRNPattern`, `TestHEALTHCAREICD10Pattern`, `TestHEALTHCAREInsuranceIDPattern`
- Cross-pattern: `TestSIREN_SSN_CrossPattern` (3 subcases)

**Integration tests** (`internal/anonymizer/*_report_test.go`):
- `TestFINANCEEUPackPipeline` — 10 cases ✓
- `TestFINANCEEUIBANCrossPatternCreditCard` ✓
- `TestHEALTHCAREPackPipeline` — 13 cases ✓
- `TestHEALTHCAREMRNCrossPatternSSN` ✓
- `TestNLPackPipeline` — 4 cases ✓
- `TestNLBSNCrossPatternSIREN` ✓
- `TestSecretsExpandedPipeline` — 43 cases ✓
- `TestSecretsPriorityOverGLOBAL` — 11 cases ✓
- `TestLoadPacksRespectsEnabledPacksOrder` ✓

### Verdict

**Ready to merge.** All review findings resolved across 4 commits. No regressions in any pack. Coverage exceeds 85% in every changed package.

---
*Final review by Claude*

**Comment by laplaque** [comment]
Closes issue #87
